lms/djangoapps/teams/csv.py

```diff
@@ -19,3 +26,234 @@ def load_team_membership_csv(course, response):
+        """
+        teamset_ids = {ts.teamset_id for ts in self.course.teams_configuration.teamsets}
+        dupe_set = set()
+        for teamset_id in self.teamset_ids:
+            if teamset_id in dupe_set:
+                self.validation_errors.append("Teamset with id " + teamset_id + " is duplicated.")
+                return False
+            dupe_set.add(teamset_id)
+            if teamset_id not in teamset_ids:
+                self.validation_errors.append("Teamset with id " + teamset_id + " does not exist.")
+                return False
+        return True
+
+    def load_user_ids_by_teamset_id(self):
+        for teamset_id in self.teamset_ids:
+            self.user_ids_by_teamset_id[teamset_id] = {
+                membership.user_id for membership in
+                CourseTeamMembership.objects.filter(
+                    team__course_id=self.course.id, team__topic_id=teamset_id
+                )
+            }
+
+    def validate_user_enrolled_in_course(self, user):
+        """
+        Invalid states:
+            user not enrolled in course
+        """
+        if not CourseEnrollment.is_enrolled(user, self.course.id):
+            self.validation_errors.append('User ' + user.username + ' is not enrolled in this course.')
+            return False
+
+        return True
+
+    def is_username_unique(self, username, usernames_found_so_far):
+        """
+        Ensures that username exists only once in an input file
+        """
+        if username in usernames_found_so_far:
+            error_message = 'Username {} was found more than once in input file.'.format(username)
+            if self.add_error_and_check_if_max_exceeded(error_message):
+                return False
+        return True
+
+    def validate_user_assignment_to_team_and_teamset(self, row):
+        """
+        Validates a user entry relative to an existing team.
+        row is a dictionary where key is column name and value is the row value
+        [andrew],masters,team1,,team3
+        [joe],masters,,team2,team3
+        """
+        user = row['user']
+        for teamset_id in self.teamset_ids:
+            team_name = row[teamset_id]
+            if not team_name:
+                continue
+            try:
+                # checks for a team inside a specific team set. This way team names can be duplicated across
+                # teamsets
+                team = self.existing_course_teams[(team_name, teamset_id)]
+            except KeyError:
+                # if a team doesn't exists, the validation doesn't apply to it.
+                all_teamset_user_ids = self.user_ids_by_teamset_id[teamset_id]
+                error_message = 'The user {0} is already a member of a team inside teamset {1} in this course.'.format(
+                    user.username, teamset_id
+                )
+                if user.id in all_teamset_user_ids and self.add_error_and_check_if_max_exceeded(error_message):
```

Comment:
In some of the methods above, you're adding a validation error directly to `self.validation_errors` (which is fine, IMO), but here, you're using this `add_error_and_check_max` helper - is there a reason to treat these cases differently?

Reply:
The times I add it to the collection are directly are "hard fail" scenarios and we don't do any more processing. When I use the helper method, if the threshold of maximum errors hasn't been met, we will continue processing - this is more of a "soft fail".

Reply:
OK, that's helpful context, thanks 👍